### DISCUSSION TITLE
mapped_column default not passed to __init__

### DISCUSSION THREAD

**Discussion by vehlwn** [sqlalchemy/sqlalchemy#11341] [Usage Questions] [answered]
I want to use mapped_column() default [parameter](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.default) to set some computed value to a column, but this seems to work only when flushing objects to a database. Column is not set to the specified value in the generated `__init__` method although documentation says otherwise:

>when used with Declarative Dataclass Mapping, indicates a default Python value that should be applied to the keyword constructor within the generated `__init__()` method.

Here is the code:

```python
from datetime import datetime, timezone
from sqlalchemy import create_engine
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    sessionmaker,
)
from sqlalchemy import select

db_path = "sqlite+pysqlite:///:memory:"
engine = create_engine(db_path, echo=True)


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"
    id: Mapped[int] = mapped_column(primary_key=True)
    created_ts: Mapped[float] = mapped_column(
        default=lambda: datetime.now(timezone.utc).timestamp()
    )

    def __repr__(self) -> str:
        return f"User(id={self.id!r}, created_ts={self.created_ts!r})"


Base.metadata.create_all(engine)
Session = sessionmaker(engine)


with Session.begin() as session:
    u1 = User()
    print("Constructed user:", u1)
    print("created_ts:", u1.created_ts)
    session.add(u1)

    for u in session.scalars(select(User)):
        print("Selected user:", u)
```

The output is:

```bash
2024-05-01 17:48:22,849 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-05-01 17:48:22,849 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user")
2024-05-01 17:48:22,849 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-05-01 17:48:22,849 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("user")
2024-05-01 17:48:22,849 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-05-01 17:48:22,850 INFO sqlalchemy.engine.Engine
CREATE TABLE user (
        id INTEGER NOT NULL,
        created_ts FLOAT NOT NULL,
        PRIMARY KEY (id)
)


2024-05-01 17:48:22,850 INFO sqlalchemy.engine.Engine [no key 0.00010s] ()
2024-05-01 17:48:22,850 INFO sqlalchemy.engine.Engine COMMIT
Constructed user: User(id=None, created_ts=None)
created_ts: None
2024-05-01 17:48:22,852 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-05-01 17:48:22,854 INFO sqlalchemy.engine.Engine INSERT INTO user (created_ts) VALUES (?)
2024-05-01 17:48:22,854 INFO sqlalchemy.engine.Engine [generated in 0.00031s] (1714571302.854252,)
2024-05-01 17:48:22,857 INFO sqlalchemy.engine.Engine SELECT user.id, user.created_ts
FROM user
2024-05-01 17:48:22,857 INFO sqlalchemy.engine.Engine [generated in 0.00029s] ()
Selected user: User(id=1, created_ts=1714571302.854252)
2024-05-01 17:48:22,857 INFO sqlalchemy.engine.Engine COMMIT
```

As you can see `u1.created_ts` is not set after User() construction but is set after INSERT. Neither `default` nor `insert_default` work for me. Am I using the `default` property wrong or is this feature not implemented? sqlalchemy 2.0.29

**Comment by zzzeek** [comment, marked as answer]
the "default" on mapped_column, when not using dataclasses, is a SQL-only INSERT default.   It's been renamed to "insert_default" to clarify this use.  

When using dataclasses, it becomes the constructor default as per the dataclasses behavior.

so you can either use the [declarative dataclass mapping](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping) feature or you'd need to create a custom `__init__` method on your class as you would use in Python normally.

**Reply by vehlwn** [reply]
Thanks. I missed the MappedAsDataclass part which forces me to specify all non-default fields in constructor. Then I'll stick to my custom `__init__` function.

**Reply by zzzeek** [reply]
you can add init=False to any fields that you dont want to be in the constructor

**Reply by vehlwn** [reply]
>init=False

That also works. Thanks.

Can `default=` be a function? I cannot make it work:
```python
class User(MappedAsDataclass, Base):
    __tablename__ = "user"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    created_ts: Mapped[float] = mapped_column(
        default=lambda: datetime.now(timezone.utc).timestamp()
    )

    def __repr__(self) -> str:
        return f"User(id={self.id!r}, created_ts={self.created_ts!r})"


Base.metadata.create_all(engine)
Session = sessionmaker(engine)


with Session.begin() as session:
    u1 = User()
    print("Constructed user:", u1)
    session.add(u1)
```
TypeError: float() argument must be a string or a real number, not 'function'

**Reply by zzzeek** [reply]
this all drives from dataclasses so see https://docs.python.org/3/library/dataclasses.html#default-factory-functions
